jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/content/ContentSourcePublisher.java

```diff
@@ -13,147 +13,311 @@
 
 package org.eclipse.jetty.io.content;
 
+import java.util.concurrent.CancellationException;
+import java.util.concurrent.CompletionException;
 import java.util.concurrent.Flow;
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.concurrent.atomic.AtomicReference;
 
 import org.eclipse.jetty.io.Content;
+import org.eclipse.jetty.util.IteratingCallback;
 import org.eclipse.jetty.util.MathUtils;
-import org.eclipse.jetty.util.thread.AutoLock;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
 
 /**
  * <p>Wraps a {@link Content.Source} as a {@link Flow.Publisher}.
  * When content is requested via {@link Flow.Subscription#request(long)}, it is
  * read from the passed {@link Content.Source} and passed to {@link Flow.Subscriber#onNext(Object)}.
  * If no content is available, then the {@link Content.Source#demand(Runnable)} method is used to
  * ultimately call {@link Flow.Subscriber#onNext(Object)} once content is available.</p>
+ * <p>{@link Content.Source} can be consumed only once and does not support multicast subscription.
+ * {@link Content.Source} will be consumed fully, otherwise will be failed in case of any errors
+ * to prevent resource leaks.</p>
  */
 public class ContentSourcePublisher implements Flow.Publisher<Content.Chunk>
 {
-    private final Content.Source content;
+    private static final Logger LOG = LoggerFactory.getLogger(ContentSourcePublisher.class);
+
+    private final AtomicReference<Content.Source> content;
 
     public ContentSourcePublisher(Content.Source content)
     {
-        this.content = content;
+        if (content == null)
```

Comment:
Better use `Objects.requireNonNull(content)`

Reply:
@lorban thank you, fixed